scripts/usb_reset.py

```diff
@@ -14,46 +14,266 @@
+            deny.write("a")
 
-    if len(sys.argv) != 2:
-        sys.exit("usage: {} device_node".format(sys.argv[0]))
+            # Only the "/dev/null" is required after CP-25630
+            whitelist = [
+                ("/dev/null", "rw")
+            ]
+            for dev in whitelist:
+                ctl = get_ctl(*dev)
+                if ctl:
+                    allow.write(ctl)
 
-    device = sys.argv[1]
-    pattern = re.compile(r"^\d+-\d+(\.\d+)*$")
-    if pattern.match(device) is None:
-        log.debug("unexpected device node: {}".format(device))
+            tasks.write(str(pid))
+
+    except IOError as e:
+        log.debug("Failed to setup cgroup: {}".format(str(e)))
         exit(1)
 
+
+def mount(source, target, fs, flags=0):
+    if ctypes.CDLL(ctypes.util.find_library("c"), use_errno=True
+                   ).mount(source, target, fs, flags) < 0:
+        # log and continue
+        log.debug("Failed to mount {} ({}) to {} with flags {}: {}".
+                  format(source, fs, target, flags,
+                         os.strerror(ctypes.get_errno())))
+
+
+def umount(target):
+    if ctypes.CDLL(ctypes.util.find_library("c"), use_errno=True
+                   ).umount(target) < 0:
+        # log and continue
+        log.debug("Failed to umount {}: {}".
+                  format(target, os.strerror(ctypes.get_errno())))
+
+
+def attach(path, dom_id, pid, reset_only):
+    # reset device
     try:
-        bus = read_int("/sys/bus/usb/devices/{}/busnum".format(device))
-        dev = read_int("/sys/bus/usb/devices/{}/devnum".format(device))
-        reset_device(bus, dev)
-    except (IOError, ValueError):
-        log.debug(traceback.format_exc())
+        with open(path, "w") as f:
+            # USBDEVFS_RESET _IO('U', 20)
+            USBDEVFS_RESET = (ord('U') << 8) | 20
+            fcntl.ioctl(f.fileno(), USBDEVFS_RESET, 0)
+    except IOError as e:
+        # log and continue
+        log.debug("Failed to reset {}: {}".format(path, str(e)))
+
+    if reset_only:
+        return
+
+    # set device file uid/gid
+    os.chown(path, pwd.getpwnam("qemu_base").pw_uid + dom_id,
+             grp.getgrnam("qemu_base").gr_gid + dom_id)
+
+    root_dir = get_root_dir(dom_id)
+    dev_dir = root_dir + "dev/"
+    if not os.path.isdir(root_dir) or not os.path.isdir(dev_dir):
+        log.debug("Error: The chroot or dev directory doesn't exist")
+        exit(1)
+
+    if not os.path.isdir(dev_dir + "bus/"):
+        # first USB device to pass-through
+        MS_BIND = 4096  # mount flags, from fs.h
+        mount("/dev", dev_dir, "", MS_BIND)
+        setup_cgroup(dom_id, pid)
+
+    sys_dir = root_dir + "sys/"
+    # sys_dir could already be mounted because of PCI pass-through
+    if not os.path.isdir(sys_dir):
+        os.mkdir(sys_dir, 0755)
+    if not os.path.isdir(sys_dir + "devices/"):
+        mount("/sys", sys_dir, "sysfs")
+
+    # add device to cgroup allow list
+    device_ctl(path, dom_id, True)
```

Comment:
`dom_id` is conventionally written as `domid`. (Throughout this file.)

Reply:
done.